<a href="https://colab.research.google.com/github/manuelagutierrezss16/Integraci-n-de-Datos-y-Prospectiva/blob/main/3_Teoria_Credibilidad.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Caso de Estudio

Una Fintech quiere estimar el error de sus datos internos asociados a las operaciones portuarias logisticas de contenedores en diferentes puertos del mundo (EVERGREEN), y quiere evaluar igualmente, utilizando la teoría de la credibilidad, la credibilidad que tiene una base de datos de operaciones externa (COSCO) de otra empresa naviera y que puede ser potencialmente integrable a la base de datos interna o de datos observados (suyos).

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns  #Las gráficas estadísticas
import matplotlib.pyplot as plt

1. Se procede con la estimación del error de los datos internos.

In [ ]:
z=2 #Indica el numero de varianzas que se van a considerar para la confiabilidad
p=0.95 # Indica el porcentaje de los datos que se encuentran a dos (z=2) varianzas

#Se procede con la carga de los datos
nxl='/content/1. RiesgoOperacional_EVERGREEEN.xlsx'
XDB=pd.read_excel(nxl)
#display(XBD)

#1 Los datos se caracterizan en riesgo
Freq=XDB.iloc[:,3];Sev=XDB.iloc[:,4]
LDA=Freq*Sev
ND=len(LDA) #Es el numero que tengo disponible

#2 Se estima el intervalo de confianza (el 95% de los datos se encuenttran entre estos)
u= np.mean(LDA); sigma=np.std(LDA)
print("el limite inferior del 95% de los datos es:", u-2*sigma)
print("el limite superior del 95% de los datos es:", u+2*sigma)

#3 El error de mis datos a una confiabilifad del 95% es:
E= z*np.sqrt((p*(1-p))/ND)
print("El error en el calculo de mis datos es: " ,E)
#a medida que los datos van en aumento el error va a ser más bajito

#4 Que pasa con el error y la media
print("El intervalo de la media con el error es:")
print("La media menos el error es:", (u-E)*1000)
print("La media más el error es:",(u+E)*1000)
print("El desfase en dolares (en cada dato) es:",(u+E)*1000-(u-E)*1000 )

#5 El porcentaje del error se puede estimar
print("El porcentaje del error es:", (E/u)*100)

el limite inferior del 95% de los datos es: -448.6790180505134
el limite superior del 95% de los datos es: 911.2122227325483
El error en el calculo de mis datos es:  0.008231667046719376
El intervalo de la media con el error es:
La media menos el error es: 231258.37067397078
La media más el error es: 231274.8340080642
El desfase en dolares (en cada dato) es: 16.463334093423327
El porcentaje del error es: 0.0035593842618837166


/usr/local/lib/python3.13/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)
/usr/local/lib/python3.13/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Conditional Formatting extension is not supported and will be removed
  warn(msg)


Todos los datos van a tener un desfase dependiendo de los datos que tenga y de
Si no confio en mis datos elegir p= 50%, oero por lo general es el 95%, para saber que tan confiables son los datos, acá se debe de tener dentro del presupuesto un descahe de 16 dolares por contenedor o el numero de eventos de riesgo por que puede aumentar los costos

"p" puede cambiar de forma independiente a las varianzas anteriores.

**Análisis de Resultados**
El error encontrado en los datos se acerca a los 16.46 dolares por evento de riesgo. Suponiendo un total de 1000 eventos de riesgo, o 1000 contenedores que se quedaron en tierra, se tiene que tener una provision adicional en el presupuesto de las pérdidad de 1000 COntenedor*16.46 USD/CONTENEDOR. De manera general, la media de la perdida por contenedor se ubica en el intervalo [231258.37067397078, 231274.8340080642]. En este sentido el porcentaje de error esta por debajo del 1% lo que es bueno de acuerdo con el número de datos disponible. Es importante que el valor de p se debe estimar en función de la calidad de los datos (100 solicitudes de crédito - 95 son correctas - 5 hay que llamar al cliente para que traiga más documentos o mejorar la información.)

**Teoría de la Credibilidad**

¿Que tanto debo confiar en los datos externos, se parecen a los míos?
Se debe de estimar las porpiedades estadisticas asociadas a la base de datos interna y base de datos externa.
Si los datos se acercan la credibilidad aumenta, que tan cercanos estan los datos externos a los datos internos, se debe saber cuantos datos hay en ambas bases de datos.

1. Cargar los datos internos y Externos

In [ ]:
#Se procede con la carga de los datos observados
nxl='/content/1. RiesgoOperacional_EVERGREEEN.xlsx'
XDB=pd.read_excel(nxl)
#display(XDB)

#1. Los datos que caracterizan el riesgo
Freq=XDB.iloc[:,3]; Sev=XDB.iloc[:,4]
LDA=Freq*Sev; ND=len(LDA); sigmao=np.std(LDA); uo=np.mean(LDA)

#2. Cargamos los datos externos
nxle='/content/3. Datos COSCO Shipping.xlsx'
XDE=pd.read_excel(nxle)
Freqe=XDE.iloc[:,3]; Seve=XDE.iloc[:,4]
LDAe=Freqe*Seve; NDe=len(LDAe);sigmae=np.std(LDAe);ue=np.mean(LDAe)

#Varianza Ponderada
EPV=(ND*sigmao**2+NDe*sigmae**2)/(ND+NDe)
print("La ponderación de las varianzas es:",EPV)

#Media Hipotetica - Ponderado de las medias
uh=(ND*uo+NDe*ue)/(ND+NDe)
print("La media ponderada de los datos es:",uh)

#La varianza de dos variables es:
VHM=(((ND)/(ND+NDe))*uo**2+((NDe)/(ND+NDe))*ue**2)-uh**2
print("La varianza de dos variables es:",VHM)
Cr=ND/(ND+(EPV/VHM))
print("La credibilidad de mis datos es:",Cr*100)
print("La credibilidad de la base de datos externos es del:",(1-Cr)*100)

**Análisis de Resultados**
La forma de interpretar la decisión del modelo está en la fórmula final:
𝐶𝑟=(𝑁𝐷_𝑜)/(𝑁𝐷_𝑜+𝐸𝑃𝑉/𝑉𝐻𝑀)
Esta fracción es una balanza entre el volumen de tu información y la relación ruido/señal:
Si el ruido supera a la señal (𝐸𝑃𝑉>𝑉𝐻𝑀): El término 𝐸𝑃𝑉/𝑉𝐻𝑀se vuelve muy grande. Esto infla el denominador y empuja tu credibilidad 𝐶𝑟hacia 0. El modelo te está diciendo que tus datos internos son demasiado volátiles o erráticos para ser confiables por sí solos, por lo que te obligará a darle más peso a la base de datos externa (𝐿𝐷𝐴_𝑒) para calcular tu riesgo.
Si la señal es clara y fuerte (𝑉𝐻𝑀>𝐸𝑃𝑉): El término 𝐸𝑃𝑉/𝑉𝐻𝑀se vuelve pequeño. El denominador se acerca al valor del numerador (𝑁𝐷_𝑜), empujando 𝐶𝑟hacia 1. El modelo determina que tus datos internos son estables y consistentes, por lo que puedes confiar plenamente en ellos sin diluirlos con datos externos.
El peso de la muestra (𝑁𝐷_𝑜): A medida que el volumen de tus observaciones internas (𝑁𝐷_𝑜) crece hacia el infinito, este número abruma al factor 𝐸𝑃𝑉/𝑉𝐻𝑀en el denominador. Matemáticamente, 𝐶𝑟tenderá a 1. Es el principio de la ley de los grandes números: si tienes una cantidad masiva de datos históricos, el modelo confiará en tu propia distribución 𝐿𝐷𝐴, asumiendo que ya contiene suficiente información para absorber cualquier volatilidad.